blog/content/post/2020-03-27-trenchboot-nlnet-lz-validation.md

```diff
@@ -0,0 +1,174 @@
+It already has every additional tool and component added, so it will be easy to
+install it.
+
+
+## Verification
+
+To create platform with DRTM enabled and properly prepared Linux kernel you must
+install following packages:
+
+- linux-kernel-trenchboot
+- initrd-trenchboot
+- GRUB-trenchboot
+- LZ-trenchboot
+
+Reboot the platform and see if NixOS boots normally. Later in the article, I
+will show you how to verify each part of above operation and hence DRTM itself.
+
+### Linux kernel and initrd
+
+Prepare system by installing following packages: `linux-kernel-trenchboot` and
+`initrd-trenchboot`. Rebuild kernel and reboot platform.
+
+```
+What to exactly verify?
+What should be/be not seen in NixOS?
+```
+
+### GRUB
+
+There are two ways to validate if GRUB will load `slaunch` module and hence run
+SKINIT and LZ (DRTM). First, during platform boot you can see if there is
+`slaunch` option available and if you can boot from it. Second, you can verify
+content of `grub.cfg` file. Both methods with possible outputs are presented
+below.
+
+#### GRUB without slaunch
+
+Prepare your system. Make sure that `GRUB-trenchboot` package is not
+installed. If it is present, uninstall it. Reboot platform and verify outputs.
+
+1. Content of grub.cfg
+
+```
+content of grub.cfg
+```
+
+2. GRUB menu in bootlog
+
+```
+GRUB menu in bootlog without slaunch
+```
+
+#### GRUB with slaunch
+
+Prepare your system. Boot to NixOS and install `GRUB-trenchboot` package. Reboot
+platform and verify outputs.
+
+1. Content of grub.cfg
+
+```
+content of grub.cfg
+```
+
+2. GRUB menu in bootlog
+
+```
+GRUB menu in bootlog with slaunch
+```
+
+### LZ
+
+Actually, there are few aspects which can be verified in LZ. We will focus on
+those two:
+- check if LZ debug option can be enabled
+- check if LZ utilizes SHA256 algorithm when using TPM2.0 module
+
+#### check if LZ debug option can be enabled
+
+Prepare your system. Boot to NixOS and install `LZ-trenchboot` package. Reboot
+platform and verify boot log.
+
+```
+LZ logs without debug
+```
+
+Boot to NixOS once again and install `LZ-trenchboot-debug` package. Reboot
+platform and verify boot log once again.
+
+```
+LZ logs with debug
+```
+
+As you can see, debug output is more verbose then previous one. However, I
+recommend to use non-debug one in normal operation.
+
+#### check if LZ utilizes SHA256 algorithm when using TPM2.0 module
+
+Prepare your system by installing `LZ-trenchboot` package and `tpm2-tools`
+package. Reboot platform and boot to NixOS. To see correctness of LZ operation
+do the following steps:
+
+1. Run `tpm2-tool`
+
+```bash
+./tpm2_command
+
+//tpm-tool output
+```
+
+2. Run `extended_all.sh` script.
+
+```bash
+./extended_all.sh <dir-to-kernel> <dir-to-initrd>
+
+//script output
+```
+
+Above script perform same operations as LZ. It measures kernel and store value
+in PCR17-18. If script's output for PCR17-18 is exactly the same as `tpm-tool`
+output then LZ use SHA256.
```

Comment:
This script does not perform the same operations as LZ. It does not modify PCRs at all, it just emulates PCR extension operation. It only simulates what should be extended into PCR17 (by SKINIT instruction, LZ and kernel), and not PCR18. Script "extends" both SHA1 and SHA256, expected result is that SHA256 values will be the same and SHA1 won't.